From 8da41c930e03a8635cbd8c89e3e591374c232c89 Mon Sep 17 00:00:00 2001
From: Jim Meyering <meyering@redhat.com>
Date: Sat, 7 May 2011 14:25:10 +0200
Subject: [PATCH 1/2] fix a bug whereby echo c|grep '[c]' would fail for any c
 in 0x80..0xff

* src/dfa.c (setbit_case_fold) [MBS_SUPPORT]: Set the bit also
when wctob returns EOF.
* NEWS (Bug fixes): Mention it.
---
 NEWS      |    6 ++++++
 src/dfa.c |    3 ++-
 2 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/NEWS b/NEWS
index 9e9974a..0cbf9ab 100644
--- a/NEWS
+++ b/NEWS
@@ -4,6 +4,11 @@
 
 ** Bug fixes
 
+   echo c|grep '[c]' would fail for any c in 0x80..0xff, and in many locales.
+   E.g., printf '\xff\n'|grep "$(printf '[\xff]')" || echo FAIL
+   would print FAIL rather than the required matching line.
+   [bug introduced in grep-2.6]
+ 
   grep --include=FILE works once again, rather than working like --exclude=FILE
   [bug introduced in grep-2.6]

diff --git a/src/dfa.c b/src/dfa.c
index f2064ed..b41cbb6 100644
--- a/src/dfa.c
+++ b/src/dfa.c
@@ -573,7 +573,8 @@ setbit_case_fold (
   else
     {
 #if MBS_SUPPORT
-      if (wctob ((unsigned char)b) == b)
+      int b2 = wctob ((unsigned char) b);
+      if (b2 == EOF || b2 == b)
 #endif
         setbit (b, c);
     }
--
1.7.5.1.299.g6e1e4
